src/compiler/scala/reflect/macros/contexts/Typers.scala

```diff
@@ -41,7 +41,14 @@ trait Typers {
 
   def inferImplicitValue(pt: Type, silent: Boolean = true, withMacrosDisabled: Boolean = false, pos: Position = enclosingPosition): Tree = {
     macroLogVerbose(s"inferring implicit value of type $pt, macros = ${!withMacrosDisabled}")
-    universe.analyzer.inferImplicit(universe.EmptyTree, pt, false, callsiteTyper.context, silent, withMacrosDisabled, pos, (pos, msg) => throw TypecheckException(pos, msg))
+    // If there is an open implicit in the context, use its tree as the one for which the implicit needs to be inserted.
+    // Otherwise, use the empty tree. This prevents false divergent implicit errors like the one in
+    // https://github.com/scala/bug/issues/10398.
+    val tree = callsiteTyper.context.openImplicits match {
+      case oi :: _ => oi.tree
+      case _ => universe.EmptyTree
+    }
+    universe.analyzer.inferImplicit(tree, pt, false, callsiteTyper.context, silent, withMacrosDisabled, pos, (pos, msg) => throw TypecheckException(pos, msg))
```

Comment:
@retronym thanks for the explanation, it was insightful and easy to follow :)

If I understood correctly, you are suggesting to disable the divergence check for every case where `tree == EmptyTree`, which amounts to every `inferImplicitValue` call from a macro, right? What would happen if the implicit truly diverges? My guess is that the compiler would give up after some open implicit limit, which would indeed result in no false positives and false negatives at the cost of some extra time and memory expanding unnecessary implicits.

With the solution in this PR I believe that there wouldn't be different behavior depending on whether the macro was called explicitly or implicitly. If a macro gets called explicitly and triggers an implicit search like the example `Derivation` macro, only the first expansion gets called with an `EmptyTree`. Subsequent macro calls must already have a non-empty `openImplicits` list, which would still provide the information to disambiguate between calls.

If we want to keep the divergence check, the tree at the head of `openImplicits` seems to be the most accurate one, as it indicates the "latest" known location that caused the macro to run. For the `None` case in the check, would `callsiteTyper.context.tree` be an acceptable alternative? From what I could gather debugging, it is always the tree that first triggered the macro (in the partest it would be, for example, `println(implicitly[Derivation[MyTC[CustomClass]]])`).

I'm basing my reasoning in the premise that for the divergence implicit to work best, the macro context should provide the most accurate tree that triggered the current implicit search, which allows the divergence checker to detect call "loops" with increasingly complex type parameters.

Once more, please let me know if I'm missing some use case! Both special-casing `EmptyTree` and using fresh names seem solutions that would work for my case, so if you think it's the best option, I'm happy to change this PR :)